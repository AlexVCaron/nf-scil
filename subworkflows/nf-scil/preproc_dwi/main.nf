include { BETCROP_FSLBETCROP as BET_PRELIM } from '../../../modules/nf-scil/betcrop/fslbetcrop/main'
include { DENOISING_MPPCA as DENOISE_DWI } from '../../../modules/nf-scil/denoising/mppca/main'
include { UTILS_EXTRACTB0 } from '../../../modules/nf-scil/utils/extractb0/main'
include { TOPUP_EDDY } from '../../../subworflows/nf-scil/topup_eddy/main'
include { BETCROP_FSLBETCROP } from '../../../modules/nf-scil/betcrop/fslbetcrop/main'
include { PREPROC_N4 as N4_DWI } from '../../../modules/nf-scil/preproc/n4/main'
include { PREPROC_NORMALIZE as NORMALIZE_DWI } from '../../../modules/nf-scil/preproc/normalize/main'
include { IMAGE_RESAMPLE as RESAMPLE_DWI } from '../../../modules/nf-scil/image/resample/main'


workflow PREPROC_DWI {

    take:
       ch_dwi     // channel: [ val(meta), [ dwi ] ]
       ch_bval_bvec // channel: [ val(meta), [ bval, bvec] ]
       ch_rev_b0    // channel: [ val(meta), [ reverse b0 ] ]

    main:

        ch_versions = Channel.empty()

        // ** Preliminary BET DWI ** : dwi//
        BET_PRELIM ( ch_dwi )
        ch_versions = ch_versions.mix(BET_PRELIM.out.versions.first())

        // ** Denoised DWI : dwi ** //
        DENOISE_DWI ( BET_PRELIM.out.image )
        ch_versions = ch_versions.mix(DENOISE_DWI.out.versions.first())

        // ** Eddy Topup DWI ** : dwi, bval, bvec, b0 (opt), rev_dwi, rev_bval, rev_bvec, rev_b0 (opt)//
        // Need to see with Arnaud for the other inputs //
        if ( ch_rev_b0 ) {
            // ** Set up input channel with reverse b0 ** //
            ch_topup_eddy = DENOISING_MPPCA.out.image.join(ch_bval_bvec).join(ch_rev_b0)
            TOPUP_EDDY ( ch_topup_eddy )
            ch_versions = ch_versions.mix(TOPUP_EDDY.out.versions.first())
        }
        else {
            // ** Set up input channel without reverse b0 ** //
            ch_eddy = DENOISING_MPPCA.out.image.join(ch_bval_bvec)
            TOPUP_EDDY ( ch_eddy )
            ch_versions = ch_versions.mix(TOPUP_EDDY.out.versions.first())
        }

        // ** Extract b0 : dwi, bval, bvec ** //
        ch_extract_b0 = TOPUP_EDDY.out.dwi_corrected.join(
                            TOPUP_EDDY.out.bval,
                            TOPUP_EDDY.out.bvec)
        UTILS_EXTRACTB0 ( ch_extract_b0 )
        ch_versions = ch_versions.mix(UTILS_EXTRACTB0.out.versions.first())

        // ** Bet-crop DWI **: dwi, bval, bvec //
        ch_betcop = TOPUP_EDDY.out.dwi_corrected.join(
                            TOPUP_EDDY.out.bval,
                            TOPUP_EDDY.out.bvec)
        BETCROP_FSLBETCROP ( ch_betcop )
        ch_versions = ch_versions.mix(BETCROP_FSLBETCROP.out.versions.first())

        // ** N4 DWI ** : dwi, b0, b0mask //
        ch_N4 = BETCROP_FSLBETCROP.out.image.join(
                            UTILS_EXTRACTB0.out.b0,
                            BETCROP_FSLBETCROP.out.mask)
        N4_DWI ( ch_N4 )
        ch_versions = ch_versions.mix(N4_DWI.out.versions.first())

        // ** Normalize DWI ** : dwi, mask, bval, bvec//
        ch_normalize = N4_DWI.out.image.join(
                            TOPUP_EDDY.out.b0_mask,
                            TOPUP_EDDY.out.bval,
                            TOPUP_EDDY.out.bvec)
        NORMALIZE_DWI ( ch_normalize )
        ch_versions = ch_versions.mix(NORMALIZE_DWI.out.versions.first())

        // ** Resampling DWI ** :  dwi, []//
        ch_resampling = NORMALIZE_DWI.out.dwi.map{[it[0], it[1],[]]}
        RESAMPLE_DWI ( ch_resampling )
        ch_versions = ch_versions.mix(RESAMPLE_DWI.out.versions.first()))

    emit:
        dwi_resample    = RESAMPLE_DWI.out.image        // channel: [ val(meta), [ dwi_resample ] ]
        bval            = TOPUP_EDDY.out.bval_corrected // channel: [ val(meta), [ bval_corrected ] ]
        bvec            = TOPUP_EDDY.out.bvec_corrected // channel: [ val(meta), [ bvec_corrected ] ]
        b0              = UTILS_EXTRACTB0.out.b0        // channel: [ val(meta), [ b0 ] ]
        b0_mask         = TOPUP_EDDY.out.out.b0_mask    // channel: [ val(meta), [ b0_mask ] ]
        versions        = ch_versions                   // channel: [ versions.yml ]
}
