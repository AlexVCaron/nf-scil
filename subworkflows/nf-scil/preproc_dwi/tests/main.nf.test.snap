{
    "preproc_dwi_rev_dwi_nob0": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,4c61c53078316c31b4d5daf446a3d6ac"
                ]
            ],
            [
                "versions.yml:md5,054ed5e3f19bcd691db58b11d47581f1",
                "versions.yml:md5,1662ab8cb093ea6d16798a8fa82639e1",
                "versions.yml:md5,411ff2a20e6cfbe3474831a860ed7282",
                "versions.yml:md5,4f7a89ba455d122508e2c3dd5553c8ca",
                "versions.yml:md5,55e94706d9e485c8a8c64ebb565bb013",
                "versions.yml:md5,962f76092c8795ce5e6d768304e2d230",
                "versions.yml:md5,977c7cd0a14188f1dfeb0f69a819b79e",
                "versions.yml:md5,a81ac95671627741bd01aef842094340",
                "versions.yml:md5,c7a4760b17ed27d9d4df828d3677acee",
                "versions.yml:md5,ea1ce12c71e2655bfff08c00836416ec",
                "versions.yml:md5,f8e51f2cdc7ca029d03c0c92b0f5cd73"
            ],
            "test_b0.nii.gz",
            "test__resampled.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test__image_boundingBox.pkl",
            "test__image_n4.nii.gz",
            "test__resampled.nii.gz",
            "test__dwi_corrected.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-29T20:39:39.008939"
    },
    "preproc_dwi_rev_b0": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,054ed5e3f19bcd691db58b11d47581f1",
                "versions.yml:md5,411ff2a20e6cfbe3474831a860ed7282",
                "versions.yml:md5,4f7a89ba455d122508e2c3dd5553c8ca",
                "versions.yml:md5,55e94706d9e485c8a8c64ebb565bb013",
                "versions.yml:md5,962f76092c8795ce5e6d768304e2d230",
                "versions.yml:md5,977c7cd0a14188f1dfeb0f69a819b79e",
                "versions.yml:md5,a81ac95671627741bd01aef842094340",
                "versions.yml:md5,c7a4760b17ed27d9d4df828d3677acee",
                "versions.yml:md5,ea1ce12c71e2655bfff08c00836416ec",
                "versions.yml:md5,f8e51f2cdc7ca029d03c0c92b0f5cd73"
            ],
            "test_b0.nii.gz",
            "test__resampled.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test__image_boundingBox.pkl",
            "test__image_n4.nii.gz",
            "test__resampled.nii.gz",
            "test__dwi_corrected.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-29T20:04:34.029359"
    },
    "preproc_dwi_all_options": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,4c61c53078316c31b4d5daf446a3d6ac"
                ]
            ],
            [
                "versions.yml:md5,1662ab8cb093ea6d16798a8fa82639e1",
                "versions.yml:md5,411ff2a20e6cfbe3474831a860ed7282",
                "versions.yml:md5,4f7a89ba455d122508e2c3dd5553c8ca",
                "versions.yml:md5,55e94706d9e485c8a8c64ebb565bb013",
                "versions.yml:md5,962f76092c8795ce5e6d768304e2d230",
                "versions.yml:md5,977c7cd0a14188f1dfeb0f69a819b79e",
                "versions.yml:md5,a81ac95671627741bd01aef842094340",
                "versions.yml:md5,c7a4760b17ed27d9d4df828d3677acee",
                "versions.yml:md5,ea1ce12c71e2655bfff08c00836416ec",
                "versions.yml:md5,f8e51f2cdc7ca029d03c0c92b0f5cd73"
            ],
            "test_b0.nii.gz",
            "test__resampled.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test__image_boundingBox.pkl",
            "test__image_n4.nii.gz",
            "test__resampled.nii.gz",
            "test__dwi_corrected.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-29T20:07:57.246937"
    }
}